src/esw/imu_driver.py

```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
+import rospy
+from sensor_msgs.msg import Imu, MagneticField, Temperature
+from geometry_msgs.msg import Quaternion, Vector3
+from std_msgs.msg import Header
+from mrover.msg import CalibrationStatus
+
+import serial
+from serial import SerialException, SerialTimeoutException
+
+
+def main():
+    # publishers for all types of IMU data, queue size is 1 to make sure we don't publish old data
+    imu_pub = rospy.Publisher("imu/data", Imu, queue_size=1)
+    mag_pub = rospy.Publisher("imu/magnetometer", MagneticField, queue_size=1)
+    temp_pub = rospy.Publisher("imu/temp", Temperature, queue_size=1)
+    calibration_pub = rospy.Publisher("imu/calibration", CalibrationStatus, queue_size=1)
+
+    # no rate used because rate is effectively set by Arduino serial publisher
+    rospy.init_node("imu_driver")
+
+    # read serial connection info and IMU frame from parameter server
+    port = rospy.get_param("imu_driver/port", "/dev/imu")
+    baud = rospy.get_param("imu_driver/baud", 115200)
+    imu_frame = rospy.get_param("imu_driver/frame_id", "imu_link")
+
+    # create serial connection with Arduino
+    ser = serial.Serial(port, baud)
+    attempts = 0
+
+    while not rospy.is_shutdown():
+
+        # try to read a line from the serial connection,
+        # if this fails 100 times in a row then end the program
+        try:
+            line = ser.readline()
+            attempts = 0
+        except SerialException:
+            attempts += 1
+            if attempts > 100:
+                raise SerialTimeoutException(f"Unable to read from serial port {port}, timed out after 100 attempts")
```

Comment:
Idk if this is the same as roserr but is there an advantage to putting the string in the exception instead of explicitly calling rospy.logerr

Reply:
I think roserror would be better since it will print big red text on the console instead of a background python exception that will only print since we have the output=screen tag in XML

Reply:
Gotcha